### PR TITLE
fix: use BTP URI format for client config

### PR DESCRIPTION
**Pull request by michielbdejong**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,7 +43,7 @@ CONNECTOR_LEDGERS={
         key: '/tmp/snakeoil.key',
         ca: '/tmp/snakeoil-ca.crt'
       },
-      "token": "shared_secret", // shared secret between server and client
+      "token": "shared_secret", // auth_token which the server expects the client to send
```
Comment (dappelt) [inline README.md:46]: @sharafian Are there still cases where `shared_secret` is literally a shared secret like it was with plugin virtual? If not, the example string `"shared_secret"` should be renamed to avoid confusion.

File: src/lib/client.js [hunk 2]
```diff
@@ -3,23 +3,27 @@
...
+    // btp+ws://auth_username:auth_token@host:port/path
+    // btp+ws://auth_username:auth_token@host/path
+    // See also: https://github.com/interledger/rfcs/pull/300
+    const parsedBtpUri = new URL(btpUri)
+    this._authUsername = parsedBtpUri.username
+    this._authToken = parsedBtpUri.password
+    parsedBtpUri.username = ''
+    parsedBtpUri.password = ''
+    // Note that setting the parsedBtpUri.protocol does not work as expected,
+    // so removing the 'btp+' prefix from the full URL here:
```
Comment (dappelt) [inline src/lib/client.js:24]: But setting the parsed username and password to `''` works? Is there a unit test validating that `this._wsUri` does not contain user/pass?

Comment (michielbdejong) [reply src/lib/client.js:24]: Yes, those work, it's just .protocol that seems to be a bit weird when you try to change its value. Test added, also.

**Review by dappelt — COMMENTED** [review]
There should be unit tests validating that 1) the URL the client connects to does not contain user/pass and 2) that the client sends user/pass in the auth subprotocol.

Before merging remember to squash into one commit.

File: src/model/rpc.js [hunk 3]
```diff
@@ -47,7 +47,7 @@ module.exports = class BtpRpc extends EventEmitter {
     this.debug = debug
   }
 
-  async addSocket (socket, authToken) {
+  async addSocket (socket, authUsername, authToken) {
```
Comment (sharafian) [inline src/model/rpc.js]: if we have 3 positional arguments we should probably change this to take an object. Or it could be `(socket, auth)` where `auth` contains `username` and `token`. That way `isClient` can be `Boolean(auth)`.

File: src/lib/client.js [hunk 4]
```diff
@@ -3,23 +3,27 @@
...
+    assert(parsedBtpUri.toString().startsWith('btp+'), 'server uri must start with "btp+"')
+    this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
```
Comment (sharafian) [inline src/lib/client.js:26]: if we just take a substring of the btp URI, it will still contain the auth details in the URI.

File: src/lib/client.js [hunk 5]
```diff
@@ -3,23 +3,27 @@
...
-  constructor ({ server, secret, plugin, insecure }) {
-    this._server = server
+  constructor ({ btpUri, plugin }) {
     this._plugin = plugin
 
-    // The server URI must follow the format: btp+wss://host:port/path
-    // See also: https://github.com/interledger/interledger/wiki/Interledger-over-BTP
-    const parsedServer = url.parse(server)
-    assert(parsedServer.protocol.startsWith('btp+'), 'server uri must start with "btp+"')
-    this._wsUri = (insecure ? 'ws://' : 'wss://') + parsedServer.host + parsedServer.path
```
Comment (sharafian) [inline src/lib/client.js]: I'm unclear on what's wrong with this way of constructing the URL. Using a substring seems like a jankier solution. Host already contains the port, and I think it's a good idea to require a flag in order to allow an insecure connection to the server. We may want to throw an error if the protocol is `btp+ws` and `insecure` is `false`, though.

File: src/lib/client.js [hunk 6]
```diff
@@ -3,23 +3,27 @@
 const assert = require('assert')
 const WebSocket = require('ws')
 const debug = require('debug')('ilp-plugin-payment-channel-framework:client')
-const url = require('url')
+const { URL } = require('url')
 
 module.exports = class BtpClient {
-  constructor ({ server, secret, plugin, insecure }) {
-    this._server = server
+  constructor ({ btpUri, plugin }) {
```
Comment (sharafian) [inline src/lib/client.js]: why are we no longer allowing the token to be passed in independently of the URI?

File: README.md [hunk 7]
```diff
@@ -71,8 +71,7 @@ CONNECTOR_LEDGERS={
     "currency": "EUR",
     "plugin": "ilp-plugin-payment-channel-framework",
     "options": {
-      "server": "btp+wss://wallet1.example:1234/example_path",
-      "token": "shared_secret"
+      "btpUri": "btp+wss://username:shared_secret@wallet1.example:1234/example_path"
```
Comment (sharafian) [inline README.md]: The reasoning to use `server` + `secret` here was that it is easier to understand next to the other plugins like XRP escrow (cc. @justmoon)

File: src/lib/client.js [hunk 8]
```diff
@@ -3,23 +3,27 @@
...
-    // See also: https://github.com/interledger/interledger/wiki/Interledger-over-BTP
-    const parsedServer = url.parse(server)
-    assert(parsedServer.protocol.startsWith('btp+'), 'server uri must start with "btp+"')
-    this._wsUri = (insecure ? 'ws://' : 'wss://') + parsedServer.host + parsedServer.path
-    this._secret = (parsedServer.auth && parsedServer.auth.split(':')[1]) || secret
-
-    if (!secret) {
-      throw new Error('secret must be provided to BTP Client in the URL or in the configuration')
-    }
+    // The BTP URI must follow one of the following formats:
```
Comment (sharafian) [inline src/lib/client.js:12]: another supported format is `btp+wss://:auth_token@host/path` or `btp+wss://host/path`

File: src/lib/plugin.js [hunk 9]
```diff
@@ -80,25 +79,10 @@ module.exports = class PluginPaymentChannel extends EventEmitter2 {
...
-        // remove token from url
-        parsed.auth = null
-        opts.server = url.format(parsed)
-      }
-
+    if (opts.btpUri) {
+      assertOptionType(opts, 'btpUri', 'string')
       this._client = new BtpClient({
-        server: opts.server,
-        secret: opts.secret,
```
Comment (sharafian) [inline src/lib/plugin.js]: we should still allow the token to be passed in separately. There may be cases (e.g. in a UI) where the server URI is shown but not the secret. If we don't allow the token to be passed in separately, the UI would have to do string manipulation to pass in authentication information. Just like in HTTP basic auth, the URI isn't the only way to tell your browser to send those credentials.

File: src/model/rpc.js [hunk 10]
```diff
@@ -75,13 +75,13 @@ module.exports = class BtpRpc extends EventEmitter {
         contentType: btpPacket.MIME_APPLICATION_OCTET_STREAM,
         data: Buffer.from([])
       }, {
-        protocolName: 'auth_token',
+        protocolName: 'auth_username',
         contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
-        data: Buffer.from(authToken, 'utf8')
+        data: Buffer.from(authUsername, 'utf8')
```
Comment (sharafian) [inline src/model/rpc.js]: should default to `''`, because the username may not be specified

**Review by sharafian — CHANGES_REQUESTED** [review]

File: src/model/rpc.js (hunk 10 above)
Comment (michielbdejong) [reply src/model/rpc.js]: if it's not specified in the URL then `parsedBtpUri.username` would be `''` so it's safe, but I'll add unit tests for both empty authUsername and empty authToken.

File: src/lib/client.js (hunk 6 above)
Comment (michielbdejong) [reply src/lib/client.js]: Because we have a BTP URI format now, right? If you specify the token separately, then either it would match the token from the btpUri, or be incorrect, and in both cases passing the token independently would be useless.

If you pass the token separately and then also remove it from your btpUri, that would mean your btpUri would by itself be incorrect, so that would also not be a very useful move?

File: src/lib/client.js (hunk 5 above)
Comment (michielbdejong) [reply src/lib/client.js]: > require a flag in order to allow an insecure connection to the server

ok. Constructing the URL from parts if overwriting functionality from the URL class, so that's not a good idea, I think. What might be an option would be to fix it upstream so that we can do:
```js
> const { URL } = require('url')
undefined
> a = new URL('btp+wss://user:pass@example.com/asdf')
URL {
  href: btp+wss://user:pass@example.com/asdf
  protocol: btp+wss:
  username: user
  password: --------
  hostname: example.com
  pathname: /asdf
}
> a.toString()
'btp+wss://user:pass@example.com/asdf'
> a.protocol = 'wss'
'wss'
> a.toString()
'btp+wss://user:pass@example.com/asdf'
> a.protocol = 'wss:'
'wss:'
> a.toString()
'btp+wss://user:pass@example.com/asdf'
>
```

File: src/lib/client.js (hunk 8 above)
Comment (michielbdejong) [reply src/lib/client.js:12]: the first one is just `auth_username === ''`, that's fine. The second one is not supported I think (but we should discuss that in https://github.com/interledger/rfcs/pull/300)

File: src/lib/client.js (hunk 4 above)
Comment (michielbdejong) [reply src/lib/client.js:26]: no, because of https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/pull/27/files/ab763b5aae14e0effaaff8d2fe9e0cc7550812a3#diff-976824b6113e95afacd333cbfab6d019R22

File: README.md (hunk 7 above)
Comment (michielbdejong) [reply README.md]: see [my comment above](https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/pull/27#discussion_r140419857)

File: src/lib/plugin.js (hunk 9 above)
Comment (michielbdejong) [reply src/lib/plugin.js]: see [my comment above](https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/pull/27)

File: src/lib/client.js (hunk 6 above)
Comment (sharafian) [reply src/lib/client.js]: I agree with you if we're saying the BTP URI is the new version of that `ilp_secret` connection string, where it's almost treated as an opaque token. But from our discussion, I thought that the BTP URI was more like a database connection string; a convenient shortcut to combine different details, but not canonical.

Comment (michielbdejong) [reply src/lib/client.js]: I think the plugin-framework constructor options will be *the* place where BTP URIs are used. If we don't use them there, then where will we?

**Comment by dappelt** [comment]
My comments are addressed.

**Review by sharafian — CHANGES_REQUESTED** [review]
LGTM except for the change from `server` to `btpUri`. @justmoon made a good point that the name of a field should be what it actually means, not the format that it's in. I can revert that change and merge this

**Review by sharafian — APPROVED** [review]
